rbc/src/main/java/com/library/rbc/service/BookService.java

```diff
@@ -34,4 +39,63 @@ public BookDto addNewBook(BookDto bookDTO) {
     bookRepository.save(savedBook);
     return bookDTO;
   }
+
+  public Page<BookDto> getBooksBy(Pageable pageable,
+      List<String> bookCategories,
+      List<String> bookStatuses) {
+    Page<BookDto> resultedList;
+    if (bookCategories == null)
+      resultedList = returnBooksByStatuses(pageable, bookStatuses);
+    else {
+      resultedList = returnBooksByCategoriesAndStatuses(pageable,bookStatuses,bookCategories);
+    }
+    if (resultedList.getTotalElements() == 0) {
+      throw new BookNotFoundException("There are no books that match your request");
+    }
+    return resultedList;
+  }
+
+  private Page<BookDto> returnBooksByStatuses (Pageable pageable, List<String> bookStatuses){
+    if(bookStatuses == null){
+      return getAllBooks(pageable);
+    }
+    List<BookStatusDto> statuses = convertStringsToBookStatusesDto(bookStatuses);
+    return bookRepository.findByBookStatusIn(pageable, statuses);
+  }
+
+  private Page<BookDto> returnBooksByCategoriesAndStatuses(Pageable pageable, List<String> bookStatuses, List<String> bookCategories){
+    List<BookCategoryDto> categories = convertStringsToBookCategoriesDto(bookCategories);
+    if (bookStatuses == null) {
+      return bookRepository.findByBookCategoriesIn(pageable, categories);
+    } else {
+      List<BookStatusDto> statuses = convertStringsToBookStatusesDto(bookStatuses);
+      return bookRepository.findByBookStatusInAndBookCategoriesIn(pageable, statuses,
+          categories);
+    }
+  }
```

Comment:
You can use a "guard clause" (if statement with early return) instead of if-else block:

```java
  private Page<BookDto> returnBooksByCategoriesAndStatuses(Pageable pageable, List<String> bookStatuses, List<String> bookCategories) {
    var categories = convertStringsToBookCategoriesDto(bookCategories);
    if (bookStatuses == null) {
      return bookRepository.findByBookCategoriesIn(pageable, categories);
    }

    var statuses = convertStringsToBookStatusesDto(bookStatuses);
    return bookRepository.findByBookStatusInAndBookCategoriesIn(pageable, statuses, categories);
  }
```